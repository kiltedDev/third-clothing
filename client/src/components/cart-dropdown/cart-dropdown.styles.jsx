import styled from 'styled-components';
import CustomButton from '../custom-button/custom-button.component';

export const CartDropdownContainer = styled.div`
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;

  @media screen and (max-width: 800px) {
    width: calc(100vw - 20px);
    right: 0px;
    height: 80vh;
    top:70px;
  }
`

export const CartDropdownItems = styled.div`
  height: 240px;
  display: flex;
  flex-direction: column;
  @media screen and (max-width: 800px) {
    height: 75vh;
  }
`

export const EmptyMessage = styled.span`
  font-size: 18px;
  margin: 50px auto;
`

export const CartDropdownButton = styled(CustomButton)`
  margin-top: auto;
`

CartDropdownButton.displayName = 'CartDropdownButton';
EmptyMessage.displayName = 'EmptyMessage';
