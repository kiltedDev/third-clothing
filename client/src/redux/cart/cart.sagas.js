import { all, call, put, takeLatest, select } from 'redux-saga/effects';

import UserActionTypes from '../user/user.types';
import CartActionTypes from './cart.types';

import { clearCart, setCartFromFirebase } from '../cart/cart.actions';
import { getUserCartRef, newUserCart } from '../../firebase/firebase.utils';
import { selectCurrentUser } from '../user/user.selectors';
import { selectCartItems } from './cart.selectors';

export function* clearCartOnSignOut() {
  yield put( clearCart() );
}

export function* onSignOutSuccess() {
  yield takeLatest( UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut )
}

export function* updateCartInFirebase() {
  const currentUser = yield select( selectCurrentUser );
  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef( currentUser.id, currentUser.activeCart );
      const cartItems = yield select( selectCartItems );
      yield cartRef.update({ cartItems });
    } catch ( error ) {
      console.log( error.message );
    }
  }
}

export function* checkCartFromFirebase({ payload: user }) {
  const cartRef = yield getUserCartRef( user.id, user.activeCart );
  const cartSnapshot = yield cartRef.get();
  yield put( setCartFromFirebase( cartSnapshot.data().cartItems ) );
}

export function* createNewUserCart() {
  const currentUser = yield select( selectCurrentUser );
  if ( currentUser ) {
    try {
      console.log(currentUser.id);
      const newCartRef = yield newUserCart( currentUser.id );
      console.log(newCartRef);
    } catch ( error ) {
      console.log( error.message )
    }
  }
  yield put( setCartFromFirebase( [] ) );

}

export function* onUserSignIn() {
  yield takeLatest(UserActionTypes.SIGN_IN_SUCCESS, checkCartFromFirebase);
}

export function* onCartChange() {
  yield takeLatest(
    [
      CartActionTypes.ADD_ITEM,
      CartActionTypes.REMOVE_ITEM,
      CartActionTypes.CLEAR_ITEM_FROM_CART
    ],
    updateCartInFirebase
  );
}

export function* onCheckOut() {
  yield takeLatest(
    [
      CartActionTypes.CHECK_OUT_CART
    ],
    createNewUserCart
  );
}

export function* cartSagas() {
  yield( all([
    call( onSignOutSuccess ),
    call( onCartChange ),
    call( onUserSignIn ),
    call( onCheckOut )
  ]))
}
