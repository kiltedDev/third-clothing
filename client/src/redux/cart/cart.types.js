export const CartActionTypes = {
  TOGGLE_CART_HIDDEN: 'TOGGLE_CART_HIDDEN',
  HIDE_CART: 'HIDE_CART',
  ADD_ITEM: 'ADD_ITEM',
  REMOVE_ITEM: 'REMOVE_ITEM',
  CLEAR_ITEM_FROM_CART: 'CLEAR_ITEM_FROM_CART',
  CLEAR_CART: 'CLEAR_CART',
  UPDATE_CART_IN_FIREBASE: 'UPDATE_CART_IN_FIREBASE',
  SET_CART_FROM_FIREBASE: 'SET_CART_FROM_FIREBASE'
};

export default CartActionTypes;
