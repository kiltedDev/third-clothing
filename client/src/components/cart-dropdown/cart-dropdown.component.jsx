import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import { toggleCartHidden } from '../../redux/cart/cart.actions';
import { selectCartItems }  from '../../redux/cart/cart.selectors';
import CustomButton from '../custom-button/custom-button.component';
import CartItem from '../cart-item/cart-item.component';

import {
  CartDropdownContainer,
  CartDropdownItems,
  EmptyMessage
} from './cart-dropdown.styles'

const CartDropdown = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer className='cart-dropdown'>
    <CartDropdownItems
      className='cart-items'
      children={
        cartItems.length
        ? cartItems.map( cartItem => (
          <CartItem
            key={ cartItem.id}
            item={ cartItem }
          />
        )) : (
          <EmptyMessage
            className='empty-message'
            children='Your cart is empty'
          />
        )
      }
    />
    <CustomButton
      onClick={() => {
        history.push( '/checkout' );
        dispatch( toggleCartHidden() );
      }}
      children='GO TO CHECKOUT'
    />
  </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
})

export default withRouter(connect(mapStateToProps)(CartDropdown));
