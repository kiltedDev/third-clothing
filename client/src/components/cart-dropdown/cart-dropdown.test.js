import React from 'react';
import { shallow } from 'enzyme';

import { CartDropdown } from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('CartDropdown', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeAll(() => {
    mockHistory = {
      push: jest.fn()
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockItems,
      history: mockHistory,
      dispatch: mockDispatch
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it('should render CartDropdown component', () => {
    expect( wrapper.debug() ).toMatchSnapshot();
  });

  it('should call history.push when button is clicked', () => {
    wrapper.find('CartDropdownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it('should render an equal number of CartItem components as the cartItems prop', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockItems.length);
  });

  it('should render EmptyMessage if cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch
    };

    const newWrapper = shallow(<CartDropdown {...mockProps} />);
    expect( newWrapper.exists('EmptyMessage') ).toBe( true );
  });
});
