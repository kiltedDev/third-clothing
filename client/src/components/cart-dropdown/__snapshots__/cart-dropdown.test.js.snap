// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown should render CartDropdown component 1`] = `
"<styled.div className=\\"cart-dropdown\\">
  <styled.div className=\\"cart-items\\">
    <Memo(CartItem) item={{...}} />
    <Memo(CartItem) item={{...}} />
    <Memo(CartItem) item={{...}} />
  </styled.div>
  <CartDropdownButton onClick={[Function: onClick]}>
    GO TO CHECKOUT
  </CartDropdownButton>
</styled.div>"
`;
